### PR TITLE
Selenoid: Puppeteer support

### PR DESCRIPTION
**Pull request by DavertMik**
## Motivation/Description of the PR

Applicable helpers:

- [ ] Webdriver
- [ ] Puppeteer
- [ ] Nightmare
- [ ] REST
- [ ] Appium
- [ ] Protractor
- [ ] TestCafe

- Description of this PR, which problem it solves
- Resolves https://github.com/Codeception/CodeceptJS/issues/2212 (if applicable).

## Type of change

- [ ] Breaking changes
- [ ] New functionality
- [ ] Bug fix
- [ ] Documentation changes/updates
- [ ] Hot fix
- [ ] Markdown files fix - not related to source code

## Checklist:

- [ ] Tests have been added
- [ ] Documentation has been added (Run `npm run docs`)
- [ ] Lint checking (Run `npm run lint`)
- [ ] Local tests are passed (Run `npm test`)

### CODE REVIEW

File: lib/plugin/selenoid.js [hunk 1]
```diff
@@ -341,6 +342,40 @@ function setOptionsForWebdriver(config) {
   }));
 }
 
+function setOptionsForPuppeteer(config) {
+  const { remote } = require('webdriverio');
```
Comment (DavertMik) [inline lib/plugin/selenoid.js]: please check that `webdriverio` is installed

```js
try {
  require.resolve('webdriverio');
} catch (err) { 
  /// write a message
}

```

File: lib/plugin/selenoid.js [hunk 2]
```diff
@@ -341,6 +342,40 @@ function setOptionsForWebdriver(config) {
...
+        },
+      });
+      Puppeteer.isRemoteBrowser = true;
+      Puppeteer.puppeteerOptions = { browserWSEndpoint: `ws://localhost:4444/devtools/${browser.sessionId}/` };
+    });
+  };
+
+  // Launch browser for every test before
+
+  event.dispatcher.on(event.test.before, lauchPuppeteer);
```
Comment (DavertMik) [inline lib/plugin/selenoid.js]: we should also close the browser explicitly on the end of a test...

```js
 event.dispatcher.on(event.test.after, () => {
   recorder.add(() => Puppeteer._stopBrowser());
});
```

File: lib/plugin/selenoid.js [hunk 3]
```diff
@@ -341,6 +342,48 @@ function setOptionsForWebdriver(config) {
...
+  event.dispatcher.on(event.test.after, () => {
+    recorder.add('Closinng browser', () => Puppeteer._stopBrowser());
+    browser.deleteSession();
+  });
+
+
+  return null;
+}
+
 function setTestConfigForWebdriver(test) {
```
Comment (PeterNgTr) [inline lib/plugin/selenoid.js]: should we have `setTestConfigForPuppeteer`?

File: lib/plugin/selenoid.js [hunk 4]
```diff
@@ -341,6 +342,48 @@ function setOptionsForWebdriver(config) {
   }));
 }
 
+function setOptionsForPuppeteer(config) {
+  const { remote } = require('webdriverio');
+  const Puppeteer = container.helpers('Puppeteer');
+  let browser = null;
+
+  const lauchPuppeteer = (test) => {
```
Comment (elukoyanov) [inline lib/plugin/selenoid.js]: maybe `launchPuppeteer`

File: lib/plugin/selenoid.js [hunk 5]
```diff
@@ -341,6 +342,48 @@ function setOptionsForWebdriver(config) {
...
+            ...config,
+            name: test.title,
+            videoName: `${test.title}.mp4`,
+            logName: `${test.title}.log`,
+          },
+        },
+      });
+      console.log(browser.sessionId);
+      Puppeteer.isRemoteBrowser = true;
+      Puppeteer.puppeteerOptions = { browserWSEndpoint: `ws://localhost:4444/devtools/${browser.sessionId}/` };
```
Comment (elukoyanov) [inline lib/plugin/selenoid.js]: this works only locally ?

File: lib/plugin/selenoid.js [hunk 6]
```diff
@@ -341,6 +342,48 @@ function setOptionsForWebdriver(config) {
...
+    });
+  };
+
+  // Launch browser for every test before
+
+  event.dispatcher.on(event.test.before, lauchPuppeteer);
```
Comment (Vorobeyko) [inline lib/plugin/selenoid.js]: We must have parameter to restart chrome.
Because on each test restarting chrome is not advisable.

**Pull request closed without merging**